File: models/note.py
# models/note.py — Pydantic schemas (request/response shapes)
# 5. What a Note looks like over the network
from datetime import datetime
from pydantic import BaseModel
from pydantic import ConfigDict

# Whats happening? : We are defining a base schema for our Note model. This schema will be used for both creating and updating notes. 
# Why is it happening?: This is done to ensure that all notes have a title, description, and optional personal field. It is also used for validation of the incoming data. 
class NoteBase(BaseModel):
    title: str
    description: str
    personal: bool | None = None


class NoteCreate(NoteBase):
    """Schema for creating a note (all required fields)."""
    pass


class NoteUpdate(BaseModel):
    """Schema for partial updates — every field is optional."""
    title: str | None = None
    description: str | None = None
    personal: bool | None = None


class NoteResponse(NoteBase):
    """Schema returned to the client — includes DB-generated fields."""
    id: int
    created_at: datetime
    updated_at: datetime

    # Pydantic v2 style: tells Pydantic to read ORM attributes, not just dicts
    model_config = ConfigDict(from_attributes=True)

File: flutter_client/lib/models/note.dart
// models/note.dart
// Mirrors FastAPI's NoteResponse, NoteCreate, and NoteUpdate schemas.

class Note {
  final int id;
  final String title;
  final String description;
  final bool? personal;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Note({
    required this.id,
    required this.title,
    required this.description,
    this.personal,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Note.fromJson(Map<String, dynamic> json) {
    return Note(
      id: json['id'] as int,
      title: json['title'] as String,
      description: json['description'] as String,
      personal: json['personal'] as bool?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  /// Returns a copy of this Note with the given fields replaced.
  Note copyWith({
    String? title,
    String? description,
    bool? personal,
  }) {
    return Note(
      id: id,
      title: title ?? this.title,
      description: description ?? this.description,
      personal: personal ?? this.personal,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}

class NoteCreate {
  final String title;
  final String description;
  final bool? personal;

  const NoteCreate({
    required this.title,
    required this.description,
    this.personal,
  });

  Map<String, dynamic> toJson() => {
        'title': title,
        'description': description,
        if (personal != null) 'personal': personal,
      };
}

class NoteUpdate {
  final String? title;
  final String? description;
  final bool? personal;

  const NoteUpdate({this.title, this.description, this.personal});

  Map<String, dynamic> toJson() => {
        if (title != null) 'title': title,
        if (description != null) 'description': description,
        if (personal != null) 'personal': personal,
      };
}
